File: 2017/pset3/kmeans.py
import pandas as pd
import numpy as np

from draw_clusters import draw_clusters


def closest_centroid(centroids, X):
    m, _ = X.shape
    return np.array([np.argmin(np.linalg.norm(centroids - X[i, :], axis=1))
                     for i in range(m)])


def recompute_centroids(X, k, clusters_index):
    return np.array([np.mean(X[clusters_index == i, :], axis=0)
                     for i in range(k)])


def k_means(X, k=3):
    m, n = X.shape
    old_centroids = np.zeros((k, n))
    index = np.random.permutation(np.arange(m))[:k]
    centroids = X[index, :]
    clusters_index = np.zeros(m)
    iteration = 0

    while np.linalg.norm(old_centroids - centroids) > 1e-15:
        old_centroids = centroids

        # for each point find closest centroid
        clusters_index = closest_centroid(centroids, X)

        # recompute centroids
        centroids = recompute_centroids(X, k, clusters_index)

        print('iteration=', iteration)
        iteration += 1

    draw_clusters(X, clusters_index, centroids, k)


if __name__ == '__main__':
    X = pd.read_csv('X.dat', sep='\s+', header=None).as_matrix()
    np.random.seed(42)
    k_means(X)

File: 2008/ps3/q4-k-means/k_means.m
function [clusters, centroids] = k_means(X, k)

%%% YOUR CODE HERE

% choose k random different Xi as initial centroids
rng(42);
m = size(X,1);
n = size(X,2);
oldcentroids = zeros(k,n);
centroids = X(randperm(size(X, 1), k), :);

while (norm(oldcentroids - centroids) > 1e-15)
    oldcentroids = centroids;
    % for each point find closest centroid
    for i = 1:m
        [M, clusters(i, 1)] = min(vecnorm(centroids - X(i, :), 2, 2));
    end

    draw_clusters(X, clusters, centroids);
    pause(1);
    
    % recompute centroids
    for i = 1:k
        centroids(i, :) = mean(X(clusters == i, :));
    end
end